File: 프로그래머스/lv3/42892. 길 찾기 게임/길 찾기 게임.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <map>
#include <iostream>

using namespace std;

int N;
vector<pair<int, int>> tree;
map<int, int> nodeX;

bool cmp(vector<int> lhs, vector<int> rhs) {
    if(lhs[1] == rhs[1]) return lhs[0] < rhs[0];
    else return lhs[1] > rhs[1];
}

void setTree(int parentIdx, int childIdx) {
    int parentX = nodeX[parentIdx];
    int childX = nodeX[childIdx];
    if(parentX > childX) {
        if(tree[parentIdx].first == -1) tree[parentIdx].first = childIdx;
        else setTree(tree[parentIdx].first, childIdx);
    }
    else { // parentX < childX
        if(tree[parentIdx].second == -1) tree[parentIdx].second = childIdx;
        else setTree(tree[parentIdx].second, childIdx);
    }
}

void preorder(vector<int> *cur, int curIdx) {
    cur->push_back(curIdx + 1);
    if(tree[curIdx].first != -1) preorder(cur, tree[curIdx].first);
    if(tree[curIdx].second != -1) preorder(cur, tree[curIdx].second);
}

void postorder(vector<int> *cur, int curIdx) {
    if(tree[curIdx].first != -1) postorder(cur, tree[curIdx].first);
    if(tree[curIdx].second != -1) postorder(cur, tree[curIdx].second);
    cur->push_back(curIdx + 1);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<vector<int>> answer;
    N = nodeinfo.size();
    for(int i = 0; i < N; i++) {
        tree.push_back({-1, -1});
        nodeinfo[i].push_back(i);
        nodeX[i] = nodeinfo[i][0];
    }
    sort(nodeinfo.begin(), nodeinfo.end(), cmp);
    
    int parentIdx = nodeinfo[0][2];
    for(int i = 1; i < N; i++) {
        int childIdx = nodeinfo[i][2];
        setTree(parentIdx, childIdx);
    }
    
    vector<int> tmp1;
    preorder(&tmp1, parentIdx);
    answer.push_back(tmp1);
    
    vector<int> tmp2;
    postorder(&tmp2, parentIdx);
    answer.push_back(tmp2);
    
    return answer;
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys

sys.setrecursionlimit(10000)
# 전위 순회, 후위 순회를 stack으로 해도, 초기 tree 구성할 때 1000을 넘겨버림
# 1000이하라 했지만, main, solution, setTree 부르면서 1000을 넘기는 듯

def setTree(parentIdx, childInfo):
    global children, nodeX
    
    childX, childY, childIdx = childInfo
    parentX = nodeX[parentIdx]
    
    childDir = 0 if childX < parentX else 1
    
    if children[parentIdx][childDir] == -1:
        children[parentIdx][childDir] = childIdx
        return
    else:
        setTree(children[parentIdx][childDir], childInfo)
        
def preOrder(rootIdx):
    global preOrderIdx
    
    stack = list()
    stack.append(rootIdx)
    
    while stack:
        curIdx = stack.pop()
    
        preOrderIdx.append(curIdx + 1)
        for childDir in reversed(range(2)):
            if children[curIdx][childDir] != -1:
                stack.append(children[curIdx][childDir])
    
def postOrder(curIdx):
    global postOrderIdx
    
    stack = list()
    stack.append(curIdx)
    
    while stack:
        curIdx = stack.pop()
        postOrderIdx.append(curIdx + 1)
    
        for childDir in range(2):
            if children[curIdx][childDir] != -1:
                stack.append(children[curIdx][childDir])
                
    postOrderIdx = postOrderIdx[::-1]

def solution(nodeinfo):
    global children, nodeX, preOrderIdx, postOrderIdx
    
    children = [[-1, -1] for _ in range(len(nodeinfo))]
    nodeX = [x for [x, _] in nodeinfo]
    nodeinfos = [[-y, x, idx] for idx, [x, y] in enumerate(nodeinfo)]
    nodeinfos.sort()
    
    rootIdx = nodeinfos[0][2]
    for my, x, idx in nodeinfos[1:]:
        y = -my
        setTree(rootIdx, (x, y, idx))
    
    answer = list()
    
    preOrderIdx = list()
    preOrder(rootIdx)
    answer.append(preOrderIdx)
    
    postOrderIdx = list()
    postOrder(rootIdx)
    answer.append(postOrderIdx)
    
    return answer